examples/jsm/csm/CSM.js

```diff
@@ -0,0 +1,338 @@
+		this.near = this.camera.near;
+		this.far = data.far || this.camera.far;
+		this.aspect = data.aspect || this.camera.aspect;
+		this.cascades = data.cascades || 3;
+		this.mode = data.mode || 'practical';
+		this.shadowMapSize = data.shadowMapSize || 2048;
+		this.shadowBias = data.shadowBias || 0.000001;
+		this.lightDirection = data.lightDirection || new THREE.Vector3( 1, - 1, 1 ).normalize();
+		this.lightIntensity = data.lightIntensity || 1;
+		this.lightNear = data.lightNear || 1;
+		this.lightFar = data.lightFar || 2000;
+		this.lightMargin = data.lightMargin || 200;
+		this.customSplitsCallback = data.customSplitsCallback;
+
+		this.lights = [];
+		this.materials = [];
+		this.createLights();
+
+		this.getBreaks();
+		this.initCascades();
+
+		this.injectInclude();
+
+	}
+
+	createLights() {
+
+		for ( let i = 0; i < this.cascades; i ++ ) {
+
+			const light = new THREE.DirectionalLight( 0xffffff, this.lightIntensity );
+			light.castShadow = true;
+			light.shadow.mapSize.width = this.shadowMapSize;
+			light.shadow.mapSize.height = this.shadowMapSize;
+
+			light.shadow.camera.near = this.lightNear;
+			light.shadow.camera.far = this.lightFar;
+			light.shadow.bias = this.shadowBias;
+
+			this.parent.add( light );
+			this.parent.add( light.target );
+			this.lights.push( light );
+
+		}
+
+	}
+
+	initCascades() {
+
+		this.mainFrustum = new Frustum( {
+			fov: this.fov,
+			near: this.near,
+			far: this.far,
+			aspect: this.aspect
+		} );
+
+		this.mainFrustum.getViewSpaceVertices();
+
+		this.frustums = this.mainFrustum.split( this.breaks );
+
+	}
+
+	getBreaks() {
+
+		this.breaks = [];
+
+		switch ( this.mode ) {
+
+			case 'uniform':
+				this.breaks = uniformSplit( this.cascades, this.near, this.far );
+				break;
+			case 'logarithmic':
+				this.breaks = logarithmicSplit( this.cascades, this.near, this.far );
+				break;
+			case 'practical':
+				this.breaks = practicalSplit( this.cascades, this.near, this.far, 0.5 );
+				break;
+			case 'custom':
+				if ( this.customSplitsCallback === undefined ) console.error( 'CSM: Custom split scheme callback not defined.' );
+				this.breaks = this.customSplitsCallback( this.cascades, this.near, this.far );
+				break;
+
+		}
+
+		function uniformSplit( amount, near, far ) {
+
+			const r = [];
+
+			for ( let i = 1; i < amount; i ++ ) {
+
+				r.push( ( near + ( far - near ) * i / amount ) / far );
+
+			}
+
+			r.push( 1 );
+			return r;
+
+		}
+
+		function logarithmicSplit( amount, near, far ) {
+
+			const r = [];
+
+			for ( let i = 1; i < amount; i ++ ) {
+
+				r.push( ( near * ( far / near ) ** ( i / amount ) ) / far );
+
+			}
+
+			r.push( 1 );
+			return r;
+
+		}
+
+		function practicalSplit( amount, near, far, lambda ) {
+
+			const log = logarithmicSplit( amount, near, far );
+			const uni = uniformSplit( amount, near, far );
+			const r = [];
+
+			for ( let i = 1; i < amount; i ++ ) {
+
+				r.push( lambda * log[ i - 1 ] + ( 1 - lambda ) * uni[ i - 1 ] );
```

Comment:
It might be more clear to use `MathUtils.lerp( unit[ i - 1 ], log[ i - 1 ], lambda )` here.

Reply:
Fixed.